### ISSUE TITLE
Compatibility with dplyr 0.6

### ISSUE THREAD

**Issue by jlegewie** [mdsumner/spdplyr#14]
[dplyr 0.6 is coming soon](https://blog.rstudio.org/2017/04/13/dplyr-0-6-0-coming-soon/). I am not sure about the implications for `spdplyr` but it's pretty consequential. `dplyr` 0.6 changes how non-standard evaluation work and as a consequence the underscored version of each main verb (filter_(), select_() etc) are no longer needed and have been deprecated. At some point, they will be removed and `spdplyr` should provide methods for `mutate` and not `mutate_`. For example, here is how `dplyr:::mutate.tbl_df` looks like in 0.6:

```
dplyr:::mutate.tbl_df <- function (.data, ...) {
    dots <- named_quos(...)
    mutate_impl(.data, dots)
}
```
So `spdplyr:::mutate.Spatial` should look something like this (I only tested this in a very limited way)

```
spdplyr:::mutate.Spatial <- function (.data, ...) {
  if (!.hasSlot(.data, "data"))
    stop("No data slot in %s", class(.data))
  dots <- dplyr:::named_quos(...)
  .data@data <- dplyr:::mutate_impl(.data@data, dots)
  .data
}
```

ps: If you update spdplyr for dplyr 0.6, the `DESCRIPTION` file should include the version of number of the `dplyr` dependence.

**Comment by mdsumner** [comment]
Thanks for this, I have already tested against old and new dplyr after some small changes and all is well. Hopefully will clean up a bit too :)

**Comment by mdsumner** [comment]
I've been though the supported verbs and I think all is well. 

Hoping to have a chance at the joins before release, appreciate any feedback now.

**Comment by jlegewie** [comment]
Yes, I will do that. Current version on github right? Anything specific that needs more testing?

**Comment by mdsumner** [comment]
Yep, all the changes are merged into the master branch. It's not passing on Appveyor but I think that's a lag problem for their systems to get rlang. I'll kick it now to see if it works

**Comment by jlegewie** [comment]
I just did some testing. Great update! Most important, I had some problems with `group_by() %>% summarize()`. In particular, two areas disappeared after using group_by with summarize. Here is the example. You can download the `SpatialPolygonsDataFrame` [here](https://www.dropbox.com/s/objj7wlo7npsap5/test.rds?dl=0). Just compare `plot(m)` with the second plot. The two areas on top in `plot(m)` disappear.

```
m <- read_rds("~/Downloads/test.rds")
test <- m %>% group_by(test_1) %>% summarize(sd = sd(test_4))
plot(m)
plot(test[1,])
plot(test[2,], add = TRUE, bor = "red")
```

Here are some other smaller things:
- `spdplyr:::arrange.Spatial` uses "to filter" in the `data_or_stop` call.
- There is no method for `right_join` and `full_join` but I guess that makes sense. The only reasonable way would be to return a `tibble` or `data.frame` discarding the `SpatialPolygonsDataFrame`.
- There are also no methods for `semi_join` and `anti_join`. They should be able to handle `SpatialPolygonsDataFrame` though.
- There are some functions in other packages from the tidyverse that work with data.frame and could be support. One example is `tidyr::unite` and `tidye::separate`

**Comment by mdsumner** [comment]
Thanks for this!   This is really helpful, and I will follow up. 

If you have some real-ish examples of the joins that would really help me, even if it's just extended descriptions. I'm keen to put some effort into thinking more about this but it's hard to get the time.  Maybe there are examples in sf we could use as a guide? 

It is possible to put the sp non-df classes in list columns, I [did this some time ago](https://github.com/r-gris/sp.df) but it didn't seem very useful and it wasn't what I really wanted at the time. It might be worth considering in the longer term. (I find spdplyr is really good for learning to code in this space, even though I don't use it directly much).

**Comment by jlegewie** [comment]
`?left_join` includes a pretty good description of all the different joins. Below is code that works with using the same SpatialPolygonsDataFrame from above.

I really like the idea of putting sp non-df classes in list columns. I actually think Hadley Wickham talked about that once (probably at a talk I attended). So in general, I think it would be great. The problem is that so many spatial packages in R use SpatialPolygonsDataFrame from sp and it's just a pain to convert...


```
m <- read_rds("~/Dropbox/test.rds")
d1 <- tibble(zone = 0:4, newvar = rnorm(5))
# left_join: return all rows from x, and all columns from x and y.
m@data %>% left_join(d1, by = "zone")
m %>% left_join(d1, by = "zone")
# All works well...

# inner_join: return all rows from x where there are matching values in y, and all columns from x and y
m@data %>% inner_join(d1, by = "zone")
m %>% inner_join(d1, by = "zone")
# All works well...

# right_join: return all rows from y, and all columns from x and y
m@data %>% right_join(d1, by = "zone")
m %>% right_join(d1, by = "zone")
# Doesn't work and it doesn't really make sense to return a `SpatialPolygonsDataFrame` here. 
# You can see that the `zome==0` row from `d1` is part of the new data.frame and there is no spatial info for that.

# semi_join: return all rows from x where there are matching values in y, keeping just columns from x.
m@data %>% semi_join(d1, by = "zone")
m %>% semi_join(d1, by = "zone")
# Doesn't work and shouldn't be a problem...

# anti_join: return all rows from x where there are not matching values in y, keeping just columns from x.
m@data %>% anti_join(d1, by = "zone")
m %>% anti_join(d1, by = "zone")
# Doesn't work and shouldn't be a problem...

# full_join: return all rows and all columns from both x and y.
m@data %>% full_join(d1, by = "zone") %>% print(n = 50)
m %>% full_join(d1, by = "zone")
# Doesn't work and it doesn't really make sense to return a `SpatialPolygonsDataFrame` here. 
# You can see that the `zome==0` row from `d1` is part of the new data.frame and there is no spatial info for that.
```

**Comment by mdsumner** [comment]
@jlegewie I released 0.1.3 already by request for the dplyr release so I didn't get back to any of this, sorry.
